PHP7.1 and PSR12 compatible code

GeoffreyDijkstra opened the pull request:
Changed code so it's PSR (12) compatible (and added github actions)
Also changed composer so it requires PHP7.1 instead of PHP7.2, code was already compatible, but composer could not require package for us, also original package required PHP ^5.6, so to also make it easier for other people to switch. Side note, because this is not breaking code, you can tag it as "hotfix" or "feature" if you want.

GeoffreyDijkstra committed:
Removed package self-reference.

GeoffreyDijkstra committed:
Composer: Added missing ext-json extension.

GeoffreyDijkstra committed:
Added github composer and phpcs validations.

IMHO the composer.json should always be valid before code gets merged to master.
For phpcs code should match PSR12 so everyone is using same code style.

GeoffreyDijkstra committed:
Reformatted code so it matches PSR12.

Code reformat so it matches PSR12 (and passes code sniffer validation)
Also few small tweaks (sprintf instead of vsprintf and 1 !== preg_match)

GeoffreyDijkstra committed:
PSR-4 complaint autoloading.

To optimize the use of composer autoloading moved everything up 2 dirs and changed composer.json to use PSR-4 autoloading.

GeoffreyDijkstra committed:
Current codebase is compatible with PHP 7.1

Because the code currently used works fine on PHP 7.1 there is no need to force people to have a higher version. Also this package replaces the old api-restclient which supported PHP ^5.6 so to make it easier for people to switch make the gap as small as possible. (ofcourse people should be encouraged to use the latest PHP version possible :wink:)

GeoffreyDijkstra committed:
Ignore vendor and composer.lock

For a composer library both the vendor folder and composer.lock can be ignored.

GeoffreyDijkstra commented:
PS: I assume (@JozefVerhoef here we go again with assuming :roll_eyes:) that you guys know that you can rebase and merge this into master so that you don't get a merge commit, also I have my fork in sync so that there are no conflicts for doing this :smile:

GeoffreyDijkstra commented:
@JozefVerhoef can you merge this? Or let me know if you don't agree with something / need something changed etc :smile:

SjonHortensius reviewed (changes requested):
Are the additional doc-comments in any way required by PSR12? I'm not sure if they are an improvement here

pbroekhof commented on composer.json:
```diff
@@ -25,22 +25,17 @@
-      "src/PostcodeNl/Api/Client.php"
-    ]
+    "psr-4": {
+      "PostcodeNl\\Api\\": "src"
```
Add a trailing slash to "src" i.e. "src/" is preferred.

GeoffreyDijkstra replied:
This is just a matter of preference it isn't required, I'll change it.

pbroekhof commented on composer.json:
```diff
@@ -25,22 +25,17 @@
+      "PostcodeNl\\Api\\": "src"
+    }
+  },
+  "replace": {
+    "postcode-nl/api-restclient": "*"
```
While PostcodeNl_Api_Client does replace api-restclient, it is not a drop-in replacement which "replace" would imply.

GeoffreyDijkstra replied:
Replace in composer just means that if you also have the original package in composer that one should not be installed. But I can leave this out if you want.

pbroekhof replied:
As far as I know, "replace" means you tell composer that this package contains the features of the original package and version, i.e. it has its classes and can be used instead of the original package. This is mainly used when you fork a package or have a bundle of multiple packages in one package. 

Which is not the case here, so it would be best not to have it.

GeoffreyDijkstra commented:
@SjonHortensius it's actually required as far as I can see https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements but at least codesniffer fails on it if you don't have them.

GeoffreyDijkstra committed:
Remove "replace" section.

Because it's not an "drop-in" replacement it's suggested to not have this included in composer.json

GeoffreyDijkstra committed:
Trailing slash in autoloader.

Trailing slash in autoloader.

pbroekhof commented:
> @SjonHortensius it's actually required as far as I can see https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements but at least codesniffer fails on it if you don't have them.

@GeoffreyDijkstra I cannot reproduce the doc-comment requirement using the phpcs.xml ruleset in the pull request. Can you supply the codesniffer error/rule that a missing doc-comment would be triggering?

GeoffreyDijkstra committed:
Made example PSR12 compliant.

Made index.php in example directory as compliant as possible with PSR12 but because it does declare multiple symbols it's excluded from phpcs checking. Also removed exclude from phpcs.xml.

GeoffreyDijkstra commented:
@pbroekhof maybe it's just me but what's the problem of having a docblocks? phpstorm generates these for me and every IDE/tool supporting dockblocks can use them. To me this point you are trying to make (which I'm missing) is not enough to block this pull request to be merged. And if you don't like something about it just checkout this branch locally make the adjustments you like and then merge it :smile: Thats the whole idea of github contributing, let somebody contribute code if it's not to your liking or standards edit it before merging (makes everybody happy) :smile:

And now to answering your question it seems this isn't a requirement anymore with latest codesniffer version (I ran codesniffer again on this project, also found that example directory was still checked so fixed that)

pbroekhof commented:
@GeoffreyDijkstra I don't mind the docblocks myself (though they can be a bit superfluous when empty), I was mainly concerned why there was a difference in PSR12 codesniffer warnings between your checkout and my own, due to your earlier comment. I wanted to resolve that before accepting the codesniffer PSR12 integration. So it wasn't really the docblocks themselves that were blocking :)

In any case, I prefer to (where possible) let the requester answer/change/tweak things first before considering taking over the pull request myself. But I think we've resolved all blocking issues now, thanks for the effort!
